[faudio] new port

rkitover opened the pull request:
Add new port FAudio, an "accuracy-focused XAudio reimplementation for
open platforms".

Use the PLATFORM_WIN32 cmake flag on Windows, otherwise use the sdl2
dependency.

MSVC build PR is here:

https://github.com/FNA-XNA/FAudio/pull/333

MSYS2 package is here:

https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-faudio/PKGBUILD

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

rkitover commented:
Please tell me what I'm doing wrong here.

rkitover commented:
I see some of the problems now, it will take me a bit more time to sort everything out.

rkitover commented:
@jimwang118 this is ready for review now, the remaining two CI problems I can do nothing about unless you have ideas. Thank you.

jimwang118 commented:
Can faudio be compiled on Android or osx? If compilation on these platforms is supported, compilation errors need to be resolved. If it is not supported, you can add `"supports": "!osx & !android"` to vcpkg.json. In addition, you need to run the command: `"./vcpkg x-add-version faudio --overwrite-version"`, and then submit the generated file.

jimwang118 commented:
I will first convert your PR to a draft, wait for you to fix all CI errors, and then click "ready for review".

rkitover commented:
If you look at the build logs for the CI failures, they are because vcpkg does not like the version string, not actual build failures. I can't do anything about that.

jimwang118 commented:
> If you look at the build logs for the CI failures, they are because vcpkg does not like the version string, not actual build failures. I can't do anything about that.

Yes, version update requires running the command: "./vcpkg x-add-version faudio --overwrite-version".

rkitover commented:
I did run that, it is in the checklist. I just ran it again too, but it is the same.

jimwang118 commented:
When I installed it on local arm64-osx, the same error message appeared. There should be a problem with the code.
```
/Users/vcpkg/jim/faudio/buildtrees/faudio/src/24.03-583b222848.clean/src/FAudio_internal_simd.c:1080:20: error: expected ';' at end of declaration
        ALIGN(int32_t, 16) data[4] =
                          ^
                          ;
/Users/vcpkg/jim/faudio/buildtrees/faudio/src/24.03-583b222848.clean/src/FAudio_internal_simd.c:1087:25: error: use of undeclared identifier 'data'; did you mean 'atan'?
        adder_frac = vld1q_s32(data);
                               ^~~~
                               atan

```

rkitover commented on ports/faudio/portfile.cmake:
```diff
@@ -0,0 +1,35 @@
+    )
+else()
+    vcpkg_cmake_configure(SOURCE_PATH "${SOURCE_PATH}")
+endif()
```
Thank you, this is fixed.

rkitover commented:
@jimwang118 I disabled clang in my `ALIGN()` macro, now those two CI checks pass, but `x86-windows` is failing for some reason, can you please find out why? The build for this triplet is succeding on my local machine.

I will check with the upstream developers to check if disabling alignment is safe to do and get back to you.

rkitover commented:
Upstream will decide on the clang issue over the next few days, I will implement it and replace the patch here.

If you would like to merge this faster, I can disable the triplets `arm64-osx` and `arm64-android` (although I need confirmation that that build environment is clang on Android.)

I believe this is the only outstanding issue.

For reference my folllowup PR and discussion is here:

https://github.com/FNA-XNA/FAudio/pull/334
.

rkitover commented:
I fixed the alignment patch, this should be ready to go now.

dg0yt commented on ports/faudio/usage:
```diff
@@ -0,0 +1,9 @@
+The package faudio provides CMake targets:
```
```suggestion
faudio provides CMake targets:
```
(Pattern given by vcpkg tool.)

dg0yt commented on ports/faudio/usage:
```diff
@@ -0,0 +1,9 @@
+    find_package(FAudio CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE FAudio::FAudio)
+
+The package faudio provides pkg-config modules:
```
```suggestion
faudio provides pkg-config modules:
```
(Pattern given by vcpkg tool.)

dg0yt commented on ports/faudio/vcpkg.json:
```diff
@@ -0,0 +1,22 @@
+  "supports": "!uwp",
+  "dependencies": [
+    {
+      "name": "sdl2",
+      "platform": "!windows"
```
```suggestion
      "name": "sdl2",
      "default-features": false,
      "platform": "!windows"
```

dg0yt reviewed (commented):
You must update `versions` after committing all changes to `ports`. (But you can amend the last commit.)

---

Unfortunately, the CMake config and the pc file don't reflect the transitive sdl2 requirement for static linkage.

`cmake/config.cmake.in` needs
~~~cmake
if(NOT "@PLATFORM_WIN32@")
   include(CMakeFindDependencyMacro)
   find_dependency(SDL2 CONFIG)
endif()
~~~
after `@PACKAGE_INIT@`.

`cmake/FAudio.pc.in` needs
~~~
Requires.private: sdl2
~~~
but only for non-win32 - maybe `sdl2` must be injected via a new variable e.g. `@PC_REQUIRES_PRIVATE@`.

dg0yt commented:
FTR the usage suggest by vcpkg's heuristic should be good enough:
~~~
faudio provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(FAudio CONFIG REQUIRED)
  target_link_libraries(main PRIVATE FAudio::FAudio)

faudio provides pkg-config modules:

    # Accuracy-focused XAudio reimplementation for open platforms
    FAudio
~~~

rkitover committed:
[faudio] new port

Add new port FAudio, an "accuracy-focused XAudio reimplementation for
open platforms".

Use the PLATFORM_WIN32 cmake flag on Windows, otherwise use the sdl2
dependency.

MSVC build PR is here:

https://github.com/FNA-XNA/FAudio/pull/333

, followup PR to fix alignment is here:

https://github.com/FNA-XNA/FAudio/pull/334

, and another followup for the SDL2 dependency:

https://github.com/FNA-XNA/FAudio/pull/335

, the MSYS2 package is here:

https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-faudio/PKGBUILD
.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

rkitover commented:
@dg0yt thank you very much for your feedback, I have fixed all the issues you flagged as you specified. Your SDL2 dependency patch was merged upstream and is included here.

jimwang118 commented:
Usage test pass with following triplets:

```
x86-windows
x64-windows
x64-windows-static
```

jimwang118 commented:
> Any outstanding issues I need to take care of or can this be merged now?

You just have to wait for it to merge.

BillyONeal committed:
Use upstream's applied patches rather than repeating them in vcpkg.

BillyONeal reviewed (approved):
I switched to a commit containing all 3 of the patches and removed them from here so it's clear we are doing exactly as upstream wishes.

Thanks for the new port!
